### PR TITLE
[Flang][OpenMP] Add host_eval clause lowering support

### PR DESCRIPTION
**Pull request by skatrak**
This patch updates Flang lowering and the `do concurrent` conversion pass to use the `host_eval` clause in `omp.target` operations to pass host information into the applicable clauses inside of the target region, instead of the previous approach where these clauses were attached to the `omp.target` operation itself.

The map info finalization pass is updated, because it previously didn't support the case where there are entry block arguments defined by other clauses placed before the map-like clauses being introduced by the pass.

The detection of when to evaluate outside of the target region is also improved for the target SPMD case in Flang lowering.

### CODE REVIEW

**Comment by skatrak** [comment]
PR stack:
- #178
- #179
- #180

**Comment by skatrak** [comment]
Ping! Can someone take a look at this and let me know if we're ok with making the transition to the community-approved approach? This PR stack is currently in the way of upstreaming target SPMD support.

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 1]
```diff
@@ -455,20 +489,23 @@ static void genNestedEvaluations(lower::AbstractConverter &converter,
...
-  if (offloadModOp.getIsTargetDevice())
+static bool mustEvalTeamsOutsideTarget(const lower::pft::Evaluation &eval,
+                                       mlir::omp::TargetOp targetOp) {
+  if (!isHostTarget(targetOp))
     return false;
 
   llvm::omp::Directive dir =
       extractOmpDirective(eval.get<parser::OpenMPConstruct>());
-  return llvm::omp::allTargetSet.test(dir) || !evalHasSiblings(eval);
+  return llvm::omp::allTeamsSet.test(dir) &&
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:499]: From the function name, it seems like `llvm::omp::allTeamsSet.test(dir)` is pre-condition (i.e. we only invoke the function from where we handle teams-related directives). Should that be an `assert`?

**Review by ergawy — APPROVED** [review]
LGTM. Just a minor comment.

**Pull request merged**